substrate/primitives/runtime/src/proving_trie/mod.rs

```diff
@@ -0,0 +1,227 @@
+			TrieError::DecodeError => "One of the proof nodes could not be decoded.",
+		}
+	}
+}
+
+/// An interface for creating, interacting with, and creating proofs in a merkle trie.
+pub trait ProvingTrie<Hashing, Key, Value>
+where
+	Self: Sized,
+	Hashing: sp_core::Hasher,
+{
+	/// Create a new instance of a `ProvingTrie` using an iterator of key/value pairs.
+	fn generate_for<I>(items: I) -> Result<Self, DispatchError>
+	where
+		I: IntoIterator<Item = (Key, Value)>;
+	/// Access the underlying trie root.
+	fn root(&self) -> &Hashing::Out;
+	/// Query a value contained within the current trie. Returns `None` if the
+	/// the value does not exist in the trie.
+	fn query(&self, key: &Key) -> Option<Value>;
+	/// Create a proof that can be used to verify a key and its value are in the trie.
+	fn create_proof(&self, key: &Key) -> Result<Vec<u8>, DispatchError>;
+	/// Verify the existence of `key` and `value` in a given trie root and proof.
+	fn verify_proof(
+		root: &Hashing::Out,
+		proof: &[u8],
+		key: &Key,
+		value: &Value,
+	) -> Result<(), DispatchError>;
+	/// This function returns the number of hashes we expect to calculate based on the
+	/// size of the proof. This is used for benchmarking, so for worst case scenario, we should
+	/// round up.
+	///
+	/// The major complexity of doing a `verify_proof` is computing the hashes needed
+	/// to calculate the merkle root. For tries, it should be easy to predict the depth
+	/// of the trie (which is equivalent to the hashes), by looking at the size of the proof.
+	/// A rough estimate should be: `proof_size` / (`hash_size` * `num_hashes_per_layer`).
+	fn proof_size_to_hashes(proof_size: &u32) -> u32;
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use crate::traits::BlakeTwo256;
+	use sp_core::H256;
+	use sp_std::collections::btree_map::BTreeMap;
+
+	// A trie which simulates a trie of accounts (u32) and balances (u128).
+	type BalanceTrie2 = base2::BasicProvingTrie<BlakeTwo256, u32, u128>;
+	type BalanceTrie16 = base16::BasicProvingTrie<BlakeTwo256, u32, u128>;
+
+	#[test]
+	fn basic_api_usage_base_2() {
+		let balance_trie = BalanceTrie2::generate_for((0..100u32).map(|i| (i, i.into()))).unwrap();
+		let root = *balance_trie.root();
+		assert_eq!(balance_trie.query(&69), Some(69));
+		assert_eq!(balance_trie.query(&6969), None);
+		let proof = balance_trie.create_proof(&69u32).unwrap();
+		assert_eq!(BalanceTrie2::verify_proof(&root, &proof, &69u32, &69u128), Ok(()));
+	}
+
+	#[test]
+	fn basic_api_usage_base_16() {
+		let balance_trie = BalanceTrie16::generate_for((0..100u32).map(|i| (i, i.into()))).unwrap();
+		let root = *balance_trie.root();
+		assert_eq!(balance_trie.query(&69), Some(69));
+		assert_eq!(balance_trie.query(&6969), None);
+		let proof = balance_trie.create_proof(&69u32).unwrap();
+		assert_eq!(BalanceTrie16::verify_proof(&root, &proof, &69u32, &69u128), Ok(()));
+	}
+
+	#[test]
+	fn proof_size_to_hashes() {
```

Comment:
Overall, the assumption here is that the bytes of the proof are mostly hashes.

Based on how many hashes are there, and the structure of the trie, we can then know how deep the trie is.

For example, if we have a base 2 trie, there should be 1 hash at for every level of the trie. For a base 16 trie, there should be 15 hashes.

Funny enough, for worst case scenario, we need `minimum_encoded_len` of the types to get a more accurate result.

To get the most accurate result, we also want to remove the key and value bytes and any other extraneous information from the proof size.

If we use `MaxEncodedLen`, we may overestimate how many bytes are actually being used for the values, and then underestimate the number of hashes we will actually do.

In this case, I assume an even worst case scenario that all bytes are hashes.

Reply:
>In this case, I assume an even worst case scenario that all bytes are hashes.

I only read the code and figured that this is what's going on. It is a pragmatic solution and i think it should be fine =)

Reply:
I made the trait more general, just `ProofToHashes`.

For Base2 trie, we can look at the number of items directly to get the depth. For Base16, we can still use this length trick.